prml/generic-asset/src/lib.rs

```diff
@@ -802,19 +820,110 @@ impl<T: Config> Module<T> {
 		AssetMeta::<T>::iter().collect()
 	}
 
-	// PRIVATE MUTABLES
+	/// Return true if the specified asset of `who` is considered dust (insignificant).
+	fn is_dust(asset_id: T::AssetId, who: &T::AccountId) -> bool {
+		let existential_deposit = AssetMeta::<T>::get(asset_id).existential_deposit();
+		// If for an asset, there is enough deposit above the defined existential deposit, it will not
+		// be considered a dust asset. Also any reservation or locks on the asset would mean the asset
+		// should be kept for the clearance of those operations and thus is not dust.
+		<FreeBalance<T>>::get(asset_id, who) < existential_deposit
+			&& <ReservedBalance<T>>::get(asset_id, who) == Zero::zero()
+			&& Self::locks(who).is_empty()
+	}
+
+	/// Update the account of `who` in the account store based on the current asset status. Pass
+	/// true in `asset_is_dust`, if the asset for `who` is at the time of this call considered dust.
+	/// This will purge dust assets.
+	fn update_account_store(asset_id: T::AssetId, who: &T::AccountId, asset_is_dust: bool) {
+		let _ = T::AccountStore::try_mutate_exists(who, |maybe_account| {
+			match maybe_account {
+				Some(account) => {
+					if asset_is_dust {
+						account.existing_assets.remove(&asset_id);
+						Self::purge(asset_id, who);
+					} else {
+						account.existing_assets.insert(asset_id);
+					}
```

Comment:
Looks like the system pallet can help (inspiration from here: https://github.com/paritytech/substrate/blob/3efe87daba09be18b3596e075c3733fe47bafdd5/frame/balances/src/lib.rs#L1074):
```suggestion
						Self::purge(asset_id, who);
						// GA may have the responsibility to keep an account 'alive'
						// if there are no consumers or another provider for the account exists, then we can release the storage safely
						if !account.should_exist() && (!frame_system::Pallet::<T>::is_provider_required(who) || frame_system::Pallet::<T>::providers(who) > One::one()) {
							*maybe_account = None
						}
					} else {
						account.existing_assets.insert(asset_id);
					}
```

Reply:
The problem was not just with removing the account (`*maybe_account = None`). Any change that would leave the account value to be the same as the default value would fail when a provider is required. That's why we need something like the flag `exists` that would be different than default for the case that we want to remove the last dust asset from existing_assets but leave the account for its consumer. The flag solution is efficient and better than coupling Generic Asset more and more with the system module.

Reply:
Makes sense. Allowing account to be reclaimed by system (i.e. nonce reaped) seems like too much complexity vs. benefit, have proposed preventing it when GA is the account store in #190. Can tackle it separately if we find the need for it in future